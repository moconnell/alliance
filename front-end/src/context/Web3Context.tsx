import React, { createContext, useContext } from "react";
import { useWeb3 } from "../hooks";
import { web3InitialState } from "../reducers";

const Web3Context = createContext(web3InitialState);

interface Web3ContextProviderProps {
  children: React.ReactNode;
}

const Web3ContextProvider = ({ children }: Web3ContextProviderProps) => {
  const web3ProviderState = useWeb3();

  return (
    <Web3Context.Provider value={web3ProviderState}>
      {children}
    </Web3Context.Provider>
  );
};

export function useWeb3Context() {
  return useContext(Web3Context);
}

export default Web3ContextProvider;
