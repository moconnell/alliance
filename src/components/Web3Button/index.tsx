import React from "react";
import { useWeb3Context } from "../../context/Web3Context";
import Button from "../Button";

interface ConnectProps {
  connect: (() => Promise<void>) | null;
}

const ConnectButton = ({ connect }: ConnectProps) => {
  return connect ? (
    <Button onClick={connect}>Connect Wallet</Button>
  ) : (
    <Button>Loading...</Button>
  );
};

interface DisconnectProps {
  title?: string;
  disconnect: (() => Promise<void>) | null;
}

const DisconnectButton = ({ title, disconnect }: DisconnectProps) => {
  return disconnect ? (
    <Button title={title} onClick={disconnect}>
      Disconnect Wallet
    </Button>
  ) : (
    <Button>Loading...</Button>
  );
};

const Web3Button = () => {
  const { address, network, web3Provider, connect, disconnect } =
    useWeb3Context();

  return web3Provider ? (
    <DisconnectButton
      disconnect={disconnect}
      title={`Disconnect ${network?.name} ${address}`}
    />
  ) : (
    <ConnectButton connect={connect} />
  );
};

export default Web3Button;
