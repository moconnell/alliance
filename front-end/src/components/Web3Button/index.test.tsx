import { render } from "@testing-library/react";
import { useWeb3Context } from "../../context/Web3Context";
import { web3InitialState } from "../../reducers";
import Web3Button from "./index";

jest.mock("../../context/Web3Context");

describe("<Web3Button />", () => {
  const mockConnect = jest.fn();
  const mockDisconnect = jest.fn();

  beforeEach(() => {
    mockConnect.mockReset();
    mockDisconnect.mockReset();
  });
  afterAll(() => {
    mockConnect.mockRestore();
    mockDisconnect.mockRestore();
  });

  const testData = [
    web3InitialState,
    { ...web3InitialState, connect: mockConnect },
    { ...web3InitialState, disconnect: mockDisconnect, web3Provider: {} },
    { ...web3InitialState, web3Provider: {} },
  ];

  testData.forEach((web3Client) =>
    it("should render correctly", async () => {
      (useWeb3Context as jest.Mock).mockReturnValue(web3Client);

      const { asFragment } = render(<Web3Button />);
      expect(asFragment()).toMatchSnapshot();
    })
  );
});

jest.clearAllMocks();
