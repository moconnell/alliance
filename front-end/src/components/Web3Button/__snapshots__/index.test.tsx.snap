// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Web3Button /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ bbkjqh"
  >
    <button
      class="sc-gsnTZi kKvmkM"
      data-testid="button"
      type="button"
    >
      Loading...
    </button>
  </div>
</DocumentFragment>
`;

exports[`<Web3Button /> should render correctly 2`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ bbkjqh"
  >
    <button
      class="sc-gsnTZi kKvmkM"
      data-testid="button"
      type="button"
    >
      Connect Wallet
    </button>
  </div>
</DocumentFragment>
`;

exports[`<Web3Button /> should render correctly 3`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ bbkjqh"
  >
    <button
      class="sc-gsnTZi kKvmkM"
      data-testid="button"
      title="Disconnect undefined null"
      type="button"
    >
      Disconnect Wallet
    </button>
  </div>
</DocumentFragment>
`;

exports[`<Web3Button /> should render correctly 4`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ bbkjqh"
  >
    <button
      class="sc-gsnTZi kKvmkM"
      data-testid="button"
      type="button"
    >
      Loading...
    </button>
  </div>
</DocumentFragment>
`;
